docs/RELEASE.md

```diff
@@ -0,0 +1,154 @@
+
+## Crate Bumping
+
+Cadence: (possibly) each Merge Request. Responsible: Developer that opened the MR.
+
+Following SemVer isn't easy, but there exists [a guide](https://doc.rust-lang.org/cargo/reference/semver.html) in the Rust documentation that explains the small details on when to bump what. This process should be augmented with CI checks that utilize [`cargo-semver-checks`](https://github.com/obi1kenobi/cargo-semver-checks) and/or [`cargo-public-api`](https://github.com/Enselic/cargo-public-api). They must also pay attention to downstream dependencies that require a version bump, because they export the changed API.
+
+### Steps
+
+1. Developer opens a Merge Request with changed crates against `master`.
+2. They bump all changed crates according to SemVer.
+3. They bump all crates that export any changed types in their *public API*.
+4. They also bump all crates that inherit logic changes from relying on one of the bumped crates. 
+
+## Stable Release
+
+Cadence: every two weeks. Responsible: Release Team.
+
+This process aims to release the `stable` branch as a *Stable* release every two weeks.
+
+### Steps
+
+1. Check if process [Clobbering](#clobbering) needs to happen and do so, if that is the case.
+2. Check out the latest commit of `stable`.
+3. Update the `CHANGELOG.md` version, date and compile the content using the prdoc files.
+4. Open a Merge Request against `stable` for visibility.
+5. Check if there were any changes since the last release and abort, if not.
+6. Run `cargo semver-checks` and `cargo public-api` again to ensure that there are no SemVer breaks.
+7. Internal QA from the release team can happen here.
+8. Do a dry-run release to ensure that it *should* work.
+10. Comment that a *Stable* release will happen from the merged commit hash.
+11. Release all changed crates to crates.io.
+12. Create a release on GitHub.
+13. Notify Devops so that they can update Westend to the new runtime.
+
+## Nightly Release
+
+Cadence: every day at 00:00 UTC+1. Responsible: Release Team
+
+This process aims to release the `master` branch as a *Nightly* release. The process can start at 00:00 UTC+1 and should automatically do the following steps.
+
+1. Check out the latest commit of branch `master`.
+2. Compare this commit to the latest `nightly*` tag and abort if there are no changes detected.
+3. Set the version of all crates to `major.0.0-nightlyYYMMDD` where `major` is the last released `major` version of that crate plus one.
+4. Tag this commit as `nightlyYYMMDD`.
+5. Do a dry-run release to ensure that it *should* work.
+6. Push this tag (the commit will not belong to any branch).
+8. Release all crates that had changed since the last nightly release to crates.io.
```

Comment:
Basically we would need to release all of the crates or not? Or the tool that checks if something changed in a crate, needs to take into consideration the dependency graph. Aka we changed something deep inside the graph and thus all crates that depend on this crate are also bumped.

Which brings up another point, we can not just do what you said in point 3. We only need to bump the version of the crates that changed. We will then also need to depend on some of the older nightly releases in crates. In the end we will end up with every crate being on a different nightly tag, depending on when it was modified the last time. I mean this works, but just highlights again that we need some tooling to ensure that we use the correct crate versions for every crate in some downstream project.

Reply:
>Basically we would need to release all of the crates or not? Or the tool that checks if something changed in a crate, needs to take into consideration the dependency graph. 

AFAIK parity-publish does exactly this.

> We will then also need to depend on some of the older nightly releases in crates. In the end we will end up with every crate being on a different nightly tag, depending on when it was modified the last time.

Ah crap. Did not think about this.

>I mean this works, but just highlights again that we need some tooling to ensure that we use the correct crate versions for every crate in some downstream project.

Yea this sounds like a nightmare. We could also create an overarching `polkadot-sdk` crate that locks in all crate versions. Otherwise some custom bumper CLI...